CDCgov/prime-reportstream
# [DO NOT MERGE] Deploy Report Stream OpenAPI docs to platform runtime

This PR:

This PR has made changes as listed below so that the platform APIs docs (OpenAPI 3.0 Specs - api.yaml - which is generated by gradle task "generateOpenApi") can be hosted on a target report stream instance: e.g. local environment, stagging environment.

Note, per team, another approach is preferred: see https://github.com/CDCgov/prime-reportstream/pull/10582
for details, where hosting api swagger ui from a "static" storage - azure storage server is preferred.
 
1. Added endpoint at <base-url>/api/docs which will serve the swagger UI of the API docs, e.g. on a local instance, the end point look like: http://localhost:7071/api/docs
2. Added swagger UI dist 5.1.0 as kotlin resources - the new API Docs end point will load web resources from there

Test Steps - Trying it out on a local setup:
1. Check out repo to local
2. Follow the README to start a local instance
3. Point browser to http://localhost:7071/api/docs, the swagger UI for platform API will be load (currently it only has end points for GET, POST, DELETE Api Keys)
4. Pick the API and fill in the required parameters and execute, and exam the results etc.
5. Screen shots of above manual testing & try out are provided in the zenhub ticket: https://app.zenhub.com/workspaces/platform-6182b02547c1130010f459db/issues/gh/cdcgov/prime-reportstream/10262
## Changes
- Actual code changes and source files additions - see Changed Files page of github
- Swagger UI dist 5.1.0 (with customizations) added as kotlin resources

## Checklist

### Testing
- [x] Manual end user test: spin up a local RS server as instructed by README, when the server is up and running and ready for API calls, point browser to http://localhost:7071/api/docs, should be able to see swagger UI with API endpoints from ApiKeysFunctions.kt available, you can call POST, GET, DELETE supplying the parameters as with postman API calls

### Process

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- Might need approval from CDC stake holder
- Before go live (e.g. deploy to STAGGING environment and make it publicly available), We can hide this by putting swagger UI behind a feature flag

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - /api/docs/[*.yaml, *.css, *.js]

- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
